packages/blocks/shuffle/src/reducer.ts

```diff
@@ -0,0 +1,93 @@
+import { Reducer } from "react";
+import produce from "immer";
+import { v4 as uuid } from "uuid";
+
+export type Item = {
+  id: string;
+  value: string;
+};
+
+export type Items = Item[];
+
+type Action<S> = {
+  type: S;
+};
+
+type ActionWithPayload<S, T> = Action<S> & {
+  payload: T;
+};
+
+export enum ActionType {
+  ADD = "add",
+  UPDATE = "update",
+  DELETE = "delete",
+  REORDER = "reorder",
+  SHUFFLE = "shuffle",
+}
+
+export type ShuffleReducerAction =
+  | ActionWithPayload<ActionType.ADD, { index: number }>
+  | ActionWithPayload<ActionType.UPDATE, { index: number; value: string }>
+  | ActionWithPayload<ActionType.DELETE, { index: number }>
+  | ActionWithPayload<
+      ActionType.REORDER,
+      { sourceIndex: number; destinationIndex: number }
+    >
+  | Action<ActionType.SHUFFLE>;
+
+export const initialItems = [
+  { id: uuid(), value: "Thing 1" },
+  { id: uuid(), value: "Thing 2" },
+];
+
+export const shuffleReducer: Reducer<Items, ShuffleReducerAction> = (
+  items,
+  action,
+) => {
+  switch (action.type) {
+    case ActionType.ADD:
+      return produce(items, (draftItems) => {
+        draftItems.splice(action.payload.index, 0, {
+          id: uuid(),
+          value: `Thing ${items.length + 1}`,
+        });
+      });
+
+    case ActionType.UPDATE:
+      return produce(items, (draftItems) => {
+        const { index, value } = action.payload;
+
+        if (draftItems[index]) {
+          draftItems[index]!.value = value;
+        }
+      });
+
+    case ActionType.DELETE:
+      return produce(items, (draftItems) => {
+        draftItems.splice(action.payload.index, 1);
+        if (draftItems.length === 0) {
+          draftItems.push({ id: uuid(), value: "Thing 1" });
+        }
```

Comment:
I would remove this – if someone wants to delete all the items, let them.

When we add the ability to choose entities from the db to load as items, people will want to be able to delete all the local / primitive values.

Reply:
I went with this as was described [here](https://www.notion.so/hashintel/PRD-Shuffle-Block-71a347c044634a958561ea2b008ae7bf#e930394410fa4023b78b3121757ba3e4), but I'm happy to change it.

One problem that would arise from changing this is that we would have no "Add Item" button anymore because they're on each item and we would have no items left, but I could move those buttons to a single general "Add Item" button at the top, next to the Shuffle Button.

Reply:
Done in 2cc9b17384f73628fcfe27161defb326fe24020c and 2cc9b17384f73628fcfe27161defb326fe24020c